### PR TITLE
nn: add DenseGeneral generalized linear layer

### PR DESCRIPTION
**Pull request by knottwill**
### Summary
Add **`torch.nn.DenseGeneral`**, a generalised fully-connected layer
that contracts *N* input axes (equivalent to `flax.linen.DenseGeneral`)
using `torch.tensordot`.

### Motivation
JAX/Flax and other ecosystems rely heavily on `DenseGeneral`
(e.g. parallel attention projections and multi-head outputs).
Adding it to core PyTorch closes a long-standing feature gap,
avoids re-implementations downstream, and lets users stay in
torchscript/compile without custom ops. A reference implementation
has been used in the wild for >1 year :contentReference[oaicite:1]{index=1}.

### What’s in this PR
* **`torch/nn/modules/dense_general.py`** – new module
* Boiler-plate imports in `torch/nn/modules/__init__.py` and `torch/nn/__init__.py`
* Docs entry

### API
```python
class torch.nn.DenseGeneral(
    in_shapes: Tuple[int, ...],
    out_features: Tuple[int, ...],
    *,
    axis: Tuple[int, ...] = (-1,),
    bias: bool = True,
    device: Optional[torch.device] = None,
    dtype: Optional[torch.dtype] = None
)

### CODE REVIEW

**Review by mikaylagawarecki — COMMENTED** [review]
Hi,  PyTorch Core has some pretty strict rules when it comes to adding a new Optimizer or Module function or feature: https://github.com/pytorch/pytorch/wiki/Developer-FAQ#i-have-a-new-function-or-feature-id-like-to-add-to-pytorch-should-it-be-in-pytorch-core-or-a-library-like-torchvision. 


It would be best to open an issue with the contents in the PR descriptions and only send a pull request if the issue is marked as actionable :)

On a separate note, I'm curious how this `DenseGeneral` layer differs from `torch.flatten`ing the last few dims of the input tensor before passing it into an `nn.Linear`?
